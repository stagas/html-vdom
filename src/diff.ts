import { VList } from './v'

// import { inspect } from 'util'
// const log = (...args: any[]) => console.log(...args.map(x => inspect(x, false, Infinity, true)))

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type ParentLike = ParentNode | VList<any>

const $ = (P: ParentLike) => (Array.isArray(P) ? P.at(-1) : P.lastChild)

// const nextSibling = <T>(parent: ParentLike<T>, child: T | ChildNode): T | ChildNode | null =>
//   parent instanceof VList ? parent.nextSiblingOf(child as T) : (child as ChildNode).nextSibling

const _ = <T>(parent: ParentLike | T[], index: number): T | null =>
  Array.isArray(parent)
    ? (parent[index] as T) ?? null
    : ((parent as ParentNode).childNodes[index] as unknown as T) ?? null

const len = (P: ParentLike) => (Array.isArray(P) ? P.length : P.childNodes.length)

export const diff = <T>(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  eq: (a: any, b: any) => boolean,
  P: ParentLike,
  b: T[],
  a: T[] = []
) => {
  // log(a, b)

  // A holds the anchor element where insertBefore(x, A) latches on
  let A: T | null

  // B holds the current element in the new list b
  let B: T | null

  // C holds a temp value of the list a that will be compared and maybe inserted
  let C: T | null

  // log(b, a)
  let i = 0

  const n = b.length
  let p = a.length

  //              p
  // a  1 2 3 4 5
  //        i
  // b  1 2 0 4 5
  //              n
  // R  1 2
  //
  // advance start pointer while equal head
  // TODO:                     set A B here
  while (i < n && i < p && eq((A = _(a, i)), (B = _(b, i)))) i++

  //          p
  // a  1 2 3 4 5
  //        i
  // b  1 2 0 4 5
  //          n
  // R  1 2
  //
  // shorten end pointer while equal tail
  // while (i < n && i < p && eq(_(a, p - 1), _(b, n - 1))) --p, --n

  // insert:
  //        A
  //        p
  // a  1 2 4 5
  //              i
  // b  1 2 0 0 0 4 5
  //              n
  // R  1 2 0 0 0
  //

  // append:
  //          A
  //          p
  // a  1 2 3
  //          i
  // b  1 2 3 4
  //            n
  //

  // prepend:
  //
  //    A
  //    p
  // a  1 2 3
  //    i
  // b  4 1 2 3
  //      n
  //

  // insert/append/prepend shortcut
  if (i == p) {
    A = _(a, p)
    for (; i < n; i++) P.insertBefore(_(b, i)!, A)
    // we are done -- rest of elements (if any) are the same
  } else {
    // replace:
    //
    //      A
    //        p
    // a  1 2 3
    //      i
    // b  1 4 3
    //        n
    //      B

    // replace + remaining tail:
    //
    //        A
    //        p
    // a  1 2 3
    //        i
    // b  1 4 5 6 3
    //            n
    // R  1 4 3
    //      B

    //      x
    //      A
    //          p
    // a  1 2 3 4
    //        i
    //        C
    //            y
    // b  1 3 2 4
    //          n
    //            B
    // R  1 3 2 4

    //      x
    //      A
    //            p
    // a  1 2 3 4
    //        C
    //        i
    //      y
    // b  1 3 4 2
    //      B
    //            n
    // R  1 3 4 2

    // moves:
    //
    //        x
    //                    p
    //        A
    // a  1 2 3 4 5 6 7 8
    //    i         y
    //    C
    // b  3 6 5 1 2 7 8 4
    //              B
    //                    n
    //
    // R  1 2 3 4 5 6 7 8
    // R  3 6 5 1 2 4 7 8
    //

    // let x = i
    // let y = i
    // for (; y < n; y++) {
    //   for (i = y; i < p; i++) {
    //     C = _(P, i)
    //     if (eq(C, B)) {
    //       P.insertBefore(C, A)
    //       break
    //     }
    //     //
    //   }

    //   while (y < n && x < p && eq((A = _(P, y)), (B = _(b, y)))) x++, y++

    //   //      A = _(P, x)
    // }

    let x = 0
    out: while (((A = _(P, i)!), (B = _(b, i)!), i++ < n)) {
      if (!eq(A, B)) {
        for (x = i; x < p; x++) {
          C = _(P, x)!
          if (eq(B, C)) {
            P.insertBefore(C, A)
            continue out
          }
        }
        for (x = i; x < n; x++) {
          C = _(b, x)
          if (eq(A, C)) {
            P.insertBefore(B, A)
            continue out
          }
        }
        if (A) P.replaceChild(B, A)
        else P.appendChild(B) //, null)
      }
    }
    p = len(P)
    i = n
    while (i++ < p) P.removeChild($(P))

    // remove tail:
    //
    //          p
    // a  1 2 3
    //      i
    // b  1 2
    //        n

    // remove head:
    //
    //      p
    // a  1 2 3
    //    i
    // b  2 3
    //    n
    // for (; i < p--; ) P.removeChild(_(P, i))
  }

  // if (P instanceof Node) {
  //   log('RESULT', P.innerHTML)
  // } else log('RESULT', [...P])
}
