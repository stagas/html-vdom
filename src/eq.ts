import { VAtom } from '.'

export const atomEq = (a: VAtom, b: VAtom) => {
  // prettier-ignore
  const eq = (
    (a != null && b != null) &&
    (
      (a === b) ||
      (
        (a[1] && b[1] && 'key' in a[1] && 'key' in b[1])
          ? a[1]!.key === b[1]!.key
          : a[0] === b[0]
      )
    )
  )
  // console.log('EQUAL', eq, a, b)
  return eq
}

export const domEq = (a: Node, b: Node) => {
  return a === b
}
