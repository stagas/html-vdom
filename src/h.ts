/**
 * JSX types
 * @private
 */

import type * as jsx from 'html-jsx'
declare module 'html-jsx' {
  // eslint-disable-next-line
  interface DOMAttributes<T> extends JSX.IntrinsicAttributes {}
}
declare global {
  namespace JSX {
    /**
     * The type returned by our `createElement` factory.
     * @private
     */
    type Element = VAtom

    interface IntrinsicElements extends jsx.IntrinsicElements {
      /**
       * This allows for any tag to be used.
       * @private
       */
      [k: string]: unknown
    }

    interface IntrinsicAttributes {
      /**
       * Pass an object that will be assigned the DOM element reference in its `current` property.
       * @private
       */
      ref?: VRef<unknown> | null

      /**
       * List index key - each item's `key` must be unique.
       * @private
       */
      key?: string | number
    }

    interface HTMLAttributes<T> extends jsx.HTMLAttributes<T> {}
    interface SVGAttributes<T> extends jsx.SVGAttributes<T> {}
    interface DOMAttributes<T> extends jsx.DOMAttributes<T> {}
  }
}

import { VAtom as VAtomCtor, VList } from './v'

export interface VRef<T> {
  current?: T | null
}

export const Fragment = Symbol()
export type JSXFragment = typeof Fragment
export type JSXKind = string | JSXFragment | JSXFunction
export type JSXProps = (Record<string, unknown> & { key?: string | number }) | null
export type JSXChild = JSXReturn[] | JSXReturn | string | number | boolean | null | undefined
export type JSXReturn = [JSXKind, JSXProps, VList<JSXChild>]
export interface JSXFunction {
  (props?: JSXProps & { children?: JSXChild[] }): JSXReturn
}
export interface JSXPragma {
  (kind: JSXKind, props: JSXProps, ...children: JSXChild[]): JSXReturn //VAtom | VList<VAtom>
}

export const h: JSXPragma = (kind, props, ...children) =>
  new VAtomCtor(kind, props, new VList(...(children.map(toAtom) as VAtom[]))) as VAtom

export type VAtom = [JSXKind, JSXProps, VList<VAtom>]

const toAtom = (child: JSXChild) => {
  if (child == null)
    //return // new VAtomCtor()
    return new VAtomCtor(<never>Comment) as VAtom

  switch (typeof child) {
    case 'boolean':
      return new VAtomCtor(<never>Comment) as VAtom
    // return false //new VAtomCtor() //return new VAtomCtor(<never>Comment)
    case 'string':
    case 'number':
      return new VAtomCtor(<never>Text, null, <never>child) as VAtom
    case 'object':
      if (Array.isArray(child) && !(child instanceof VAtomCtor))
        return h(Fragment, null, ...(child as JSXReturn[]))
    //Array === child?.constructor) return h(Fragment, null, ...(child as JSXReturn[]))
  }

  return child
}
