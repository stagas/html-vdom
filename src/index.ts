export * from './h'
export * from './render'
