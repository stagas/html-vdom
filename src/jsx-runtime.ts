import { bindAll, EventEmitter } from 'everyday-utils'
import type * as jsxi from 'html-jsx'

export * from 'html-jsx'

import { createProps, updateProps } from './props'
export { createProps, updateProps }

declare module 'html-jsx' {
  // eslint-disable-next-line
  interface DOMAttributes<T> extends JSX.IntrinsicAttributes { }
}

/**
 * JSX types
 * @private
 */
declare global {
  namespace JSX {
    /**
     * The type returned by our `jsx` factory.
     * @private
     */
    type Element = VKid

    interface IntrinsicElements extends jsxi.IntrinsicElements {
      /**
       * This allows for any tag to be used.
       * @private
       */
      [k: string]: unknown
    }

    interface IntrinsicAttributes {
      /**
       * Pass an object that will be assigned the DOM element reference in its `current` property.
       * @private
       */
      ref?: VRef<unknown> | null

      /**
       * List index key - each item's `key` must be unique.
       * @private
       */
      key?: object | symbol | string | number

      /**
       * Custom hook for when element is created.
       */
      onref?: (el: any) => (() => void) | void

      /**
       * Custom hook for when element is removed.
       */
      onunref?: (el: any) => (() => void) | void

      /**
       * Children.
       * @private
       */
      children?: JSX.Element

      class?: string
      className?: string
      style?: Partial<CSSStyleDeclaration> | string | false | null | void
      part?: string

      /**
       * Sets the `innerHTML` of an element to the **exact** string **without** escaping.
       */
      innerHTML?: string
    }

    interface HTMLAttributes<T> extends jsxi.HTMLAttributes<T> { }
    interface SVGAttributes<T> extends jsxi.SVGAttributes<T> { }
    interface DOMAttributes<T> extends jsxi.DOMAttributes<T> {
    }
  }
}

export type $<T> = (props: T & { children?: any }) => JSX.Element

export interface VRef<T> {
  current?: T | null | void
}

type Fn = (...args: any) => any
export type Doc = (tag: string, opts?: ElementCreationOptions) => Element
export type VFn = (props: any) => VKid
export type VKid = VKids | VNode<any> | string | number | boolean | null | undefined | void
type VKids =
  & VKid[]
  & Partial<{
    running: boolean
    dom: El[]
    flatDom: DomEl[]
    keyed: Map<string, number>
    mapped: Map<El, VKid>
  }>
type DomEl = Element | CharacterData | ChildNode
type El = DomEl | Chunk
type TargetEl = El | DocumentFragment
type VAny = VNode<any>

export type Hook = Fn & {
  fn: Fn
  onremove?: Fn
  onstart?: Fn
  onend?: Fn
} & EventEmitter<{
  remove: () => void,
  start: () => void,
  end: () => void
}>
  & Record<string, any>

export type Props = Record<string, any>

type VNode<T extends string | symbol | typeof Text | typeof Comment | VFn> = {
  kind: T
  props: Props
  key?: string
  hook?: Hook | undefined
  keep?: boolean
  onunref?: () => void
}

export const Fragment = Symbol()
export const jsx = (kind: any, props: any, key: any) =>
  kind === Fragment
    ? props.children as VKid
    : { kind, props, key } as VNode<typeof kind>
export const jsxs = jsx

export let hook: Hook
export const createHook = () =>
  bindAll(new EventEmitter(), function current(fn: Fn = (current as Hook).fn) {
    const prev = hook
    hook = current as Hook
    hook.fn = fn

    hook.onstart?.()
    hook.emit('start')
    fn()
    hook.onend?.()
    hook.emit('end')

    hook = prev
  }) as Hook

class Chunk extends Array {
  firstChild?: any
  dom: El[] = []
  get last() {
    return this.dom.at(-1)
  }
  get nextSibling(): El | null | undefined {
    return this.last?.nextSibling
  }
  // insertBefore(x: any) {
  //   this.unshift(x)
  // }
  appendChild(x: any) {
    this.push(x)
  }
  after(x: Node) {
    this.last?.after(x)
  }
  save() {
    this.dom = [...this]
  }
  remove() {
    this.dom.forEach(el => {
      ; (el as any).onunref?.()
      el.remove()

      // if ((el as any)?.ref?.current) {
      // console.log('removing', el)
      // ; (el as any).ref.current = null
      // }

    })
    this.splice(0)
  }
  removeChild(x: any) {
    const i = this.indexOf(x)
    ~i && this.splice(i, 1)
    // i = this.dom.indexOf(x)
    // ~i && this.dom.splice(i, 1)
  }
}

export type { Chunk, VNode }

const { TEXT_NODE, COMMENT_NODE } = document
export const html = document.createElement.bind(document)
export const svg = document.createElementNS.bind(document, 'http://www.w3.org/2000/svg')
const forceArray = <T>(x: any, withNull: boolean): T =>
  (Array.isArray(x) ? (withNull && !x.length ? [null] : x) : x == null && !withNull ? [] : [x]) as unknown as T
const flatDom = (arr: El[], res: DomEl[] = []) => {
  for (const el of arr) {
    if ((el as Chunk).dom) res.push(...flatDom((el as Chunk).dom!))
    else res.push(el as DomEl)
  }
  return res
}

export const renderCache = new WeakMap()
export function render(n: VKid): DocumentFragment
export function render<T extends TargetEl>(n: VKid, el: T, doc?: Doc, withNull?: boolean): T
export function render(n: VKid, el: DocumentFragment = document.createDocumentFragment(), doc: Doc = html, withNull = false) {
  reconcileVDom(el, forceArray(n, withNull), renderCache.get(el), doc)
  return el
}

function reconcileVDom(parent: TargetEl, nk: VKids, pk: VKids | VKid, doc: Doc) {
  if ((pk as VKids)?.running) {
    console.warn('attempt to reconcile lane which is reconciling')
    return
  }

  if (pk === nk) {
    // same?
    // return
    nk = [...nk]
  }

  // ?
  nk.running = true

  renderCache.set(parent, nk)

  // for (const [i, n] of nk.entries() as any) {
  //   nk[i] = n?.valueOf?.()
  // }

  nk.dom = Array(nk.length)
  nk.keyed = new Map()
  nk.mapped = new Map()

  if (Array.isArray(pk)) {
    const keep = new Set()
    for (let i = 0, n, el, p, pel, k, pi; i < nk.length; i++) {
      n = nk[pi = i]
      k = (n as VAny)?.key ?? (n as any)?.ref?.key
      if (k != null) {
        nk.keyed.set(k, i)
        pi = pk.keyed!.get(k) ?? -1
      }
      p = pk[pi]
      pel = pk.dom![pi]
      nk.dom[i] = el = reconcileNode(doc, n, p, pel)
      nk.mapped.set(el, n)
      if (el === pel)
        keep.add(pel)
    }
    for (const pel of pk.dom!) {
      if (!keep.has(pel)) {
        ; (pel as any).onunref?.()
        // if ((pel as any)?.ref?.current) {
        // console.log('removing', pel)
        // ; (pel as any).ref.current = null
        // }
        // TODO: this line commented is breaking things if enabled
        // and other things break if disabled?
        // if (parent instanceof Chunk) parent.removeChild(pel)
        pel.remove()
        const hook = (pk.mapped!.get(pel) as VNode<VFn>)?.hook
        hook?.onremove?.()
        hook?.emit('remove')
        // render(null, pel)
      }
    }
  } else {
    for (let i = 0, n, el, k; i < nk.length; i++) {
      n = nk[i]
      k = (n as VAny)?.key ?? (n as any)?.ref?.key
      if (k != null)
        nk.keyed.set(k, i)
      nk.dom[i] = el = reconcileNode(doc, n)
      nk.mapped.set(el, n)
    }
  }

  nk.flatDom = flatDom(nk.dom)
  if ((pk as VKids)?.flatDom)
    diffDom(parent, nk.flatDom, (pk as VKids)!.flatDom!)

  else
    nk.flatDom.forEach(el => parent.appendChild(el))

  nk.running = false
}

// scheduling to prevent triggering ref effects before this render finishes
// TODO: has to be put in a proper queue instead of relying on the microtask
function mountEl(el: DomEl & { ref: VRef<any>; onref: any; onunref?: any }) {
  if (el?.ref && el !== el.ref.current) {
    queueMicrotask(() => {
      if (el.isConnected) {
        el.ref.current = el
      }
    })
  }
  if (el?.onref) {
    queueMicrotask(() => {
      if (el.isConnected) {
        el.onunref = el.onref(el)
      }
    })
  }

  return el
}

function diffDom(parent: TargetEl, n: DomEl[], p: DomEl[], i = 0, len = n.length, el?: DomEl, last?: DomEl) {
  if (parent instanceof Chunk) {
    // parent.splice(0) // , parent.length, ...n)
    // TODO: optimize this
    for (; i < len; i++) {
      el = n[i]
      if (i < parent.length) {
        if (p[i] === el)
          continue
        parent[i] = el
      } else {
        parent.push(el)
      }
    }
    let d = parent.length - len
    while (d--)
      parent.pop()
  } else {
    for (; i < len; i++) {
      el = n[i]
      if (p[i] === el)
        last = el
      else if (!i)
        parent.insertBefore(last = el, parent.firstChild)
      else
        last!.after(last = el)
    }
  }
}

function reconcileNode(doc: Doc, n: VKid, p?: VKid, pel?: El | null) {
  let el: El | null
  switch (typeof n) {
    case 'string':
    case 'number':
      if ((pel as Node)?.nodeType === TEXT_NODE) {
        if (p != n)
          (pel as Node).nodeValue = n as string
        return pel as El
      }
      el = new Text(n as string)
      return el
    case 'object':
      if (!n)
        break
      if (Array.isArray(n)) {
        if (pel && Array.isArray(p))
          el = pel as Chunk
        else
          el = new Chunk()
        if (!n.length)
          n.push(null)
        reconcileVDom(el, n, p, doc)
        el.save()
      } else if (typeof n.kind === 'string') {
        // maybe switch to svg namespace
        if (n.kind === 'svg')
          doc = svg
        if (( // if we pass explicit element reference, and is the right tag, use that
          n.props.ref?.current
          && n.props.ref.current.tagName.toLowerCase() === n.kind
          && (el = n.props.ref.current as El)
        ) || (
            // if the previous element is of the same kind, use that
            pel
            && (p as VNode<string>)?.kind === n.kind
            && (el = pel)
          )) {
          // only update props if we have the element
          ; (el as any).onprops?.(n.props)
          updateProps(doc, el as Element, n.kind, n.props)
        } else {
          // otherwise create an element and props
          el = doc(n.kind, 'is' in n.props ? { is: n.props.is } : void 0); (el as any).onprops?.(n.props)
          createProps(doc, el as Element, n.kind, n.props)
        }
        // maybe switch to xhtml namespace
        if (n.kind === 'foreignObject')
          doc = html
        // render children
        if (!n.kind.includes('-'))
          render(n.props.children, el, doc)
        mountEl(el as any)
      } else {
        let initial = true
        if (!((el = pel!) && (n.hook = (p as VNode<VFn>)?.hook))) {
          el = new Chunk()
          n.hook = createHook()
        }
        const anchor = new Comment()
        let prevDom: DomEl[]
        let nextDom: DomEl[]
        n.hook(function hookRun() {
          let next!: ChildNode | null
          if (!initial && !(next = el!.nextSibling as ChildNode))
            el!.after(next = anchor)
          if (typeof n.kind !== 'function') {
            console.warn('Hook called but node is not a function component.')
            console.warn(n)
            return
          }
          render(n.kind(n.props), el!, doc, true); (el as Chunk).save()
          if (!initial && next) {
            nextDom = flatDom(el as Chunk)
            if (prevDom?.length > 0) {
              for (let i = 0, last, e: El; i < nextDom.length; i++) {
                e = nextDom[i]
                if (last) {
                  last.after(last = e)
                } else if (prevDom[i] !== e) {
                  next.before(last = e)
                } else {
                  last = last && e
                }
              }
            } else {
              for (const e of nextDom)
                next.before(e)
            }
            prevDom = nextDom
            next === anchor && next.remove()
          } else {
            prevDom = flatDom(el as Chunk)
            initial = false
          }
        })
      }
      return el
  }
  if ((pel as Node)?.nodeType === COMMENT_NODE)
    return pel as El
  el = new Comment()
  return el
}
