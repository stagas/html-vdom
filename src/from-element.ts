import * as jsxi from 'html-jsx'
import { jsx } from './jsx-runtime'

export type Constructor<T> = new(...args: any[]) => T

const Registry: WeakMap<typeof HTMLElement, any> = new WeakMap()

export interface Options<T> extends ElementDefinitionOptions {
  interface?: Constructor<T>
}

export type Component<T, I> = (props: Partial<Omit<T & I, keyof HTMLElement>> & jsxi.HTMLAttributes<T>) => any

export const fromElement = <T extends HTMLElement, I = HTMLElement>(
  Element: Constructor<T>,
  options: Options<I> = {},
): Component<T, I> => {
  let fn: Component<T, I> = Registry.get(Element)
  if (fn) return fn

  const tag = 'x-' + Element.name.toLowerCase() + ((Math.random() * 1e7) | 0).toString(36)
  const ctor = class extends Element {}
  customElements.define(tag, ctor, options)
  if (options.extends)
    fn = props => jsx(options.extends, Object.assign(props, { is: tag }), void 0)
  else
    fn = props => jsx(tag, props, void 0)
  Registry.set(Element, fn)
  ;(fn as any).toString = () => tag

  return fn
}
