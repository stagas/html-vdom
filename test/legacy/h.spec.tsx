import { Fragment, Hook, current, h, render } from '../../src'

let c: HTMLDivElement
beforeEach(() => (c = document.createElement('div')))

const clean = (s: string) =>
  s.replaceAll('<!--[-->', '').replaceAll('<!--]-->', '').replaceAll('<!---->', '')
const html = () => clean(c.innerHTML)

describe('render', () => {
  describe('create', () => {
    it('Element', () => {
      render(<div></div>, c)
      expect(html()).toMatchSnapshot()
    })

    it('attribute: id', () => {
      render(<div id="foo"></div>, c)
      expect(html()).toMatchSnapshot()
    })

    it('Element with text child', () => {
      render(<div>hello</div>, c)
      expect(html()).toMatchSnapshot()
    })

    it('Element with number child', () => {
      render(<div>{123}</div>, c)
      expect(html()).toMatchSnapshot()
    })

    it('Fragment', () => {
      render(<>hello world</>, c)
      expect(html()).toMatchSnapshot()
    })

    it('Function', () => {
      const Foo = () => <>hello</>
      render(<Foo />, c)
      expect(html()).toMatchSnapshot()
    })

    it('Function returning Function', () => {
      const Bar = () => <>bar</>
      const Foo = () => <Bar />
      render(<Foo />, c)
      expect(html()).toMatchSnapshot()
    })
  })
  describe('update', () => {
    it('Element', () => {
      render(<div></div>, c)
      expect(html()).toMatchSnapshot()
      render(<p></p>, c)
      expect(html()).toMatchSnapshot()
    })

    it('Element with text child', () => {
      render(<div>hello</div>, c)
      expect(html()).toMatchSnapshot()
      render(<div>world</div>, c)
      expect(html()).toMatchSnapshot()
    })

    it('Element with number child', () => {
      render(<div>{123}</div>, c)
      expect(html()).toMatchSnapshot()
      render(<div>{456}</div>, c)
      expect(html()).toMatchSnapshot()
    })

    it('Fragment', () => {
      render(<>hello world</>, c)
      expect(html()).toMatchSnapshot()
      render(<>something else</>, c)
      expect(html()).toMatchSnapshot()
    })

    it('Function', () => {
      let i = 0
      const Foo = () => <>{i++}</>
      render(<Foo />, c)
      expect(html()).toMatchSnapshot()
      render(<Foo />, c)
      expect(html()).toMatchSnapshot()
    })

    it('Function returning Function', () => {
      let i = 0
      const Bar = () => <>{i++}</>
      const Foo = () => <Bar />
      render(<Foo />, c)
      expect(html()).toMatchSnapshot()
      render(<Foo />, c)
      expect(html()).toMatchSnapshot()
    })

    it('only updates what changed', () => {
      let i = 0
      const Bar = () => <p>{i++}</p>
      const Foo = () => (
        <div>
          <Bar />
        </div>
      )
      render(<Foo />, c)
      expect(html()).toMatchSnapshot()
      // expect(c.innerHTML).toMatchSnapshot()
      const div = c.querySelector('div')
      const p = c.querySelector('p')
      render(<Foo />, c)
      expect(html()).toMatchSnapshot()
      expect(c.querySelector('div')).toBe(div)
      expect(c.querySelector('p')).toBe(p)
      // expect(c.innerHTML).toMatchSnapshot()
    })
  })

  describe('hook', () => {
    it('updates', () => {
      let i = 0
      let hook!: Hook
      const Foo = () => {
        hook = current.hook!
        return <>{i++}</>
      }
      render(<Foo />, c)
      expect(html()).toMatchSnapshot()
      hook.trigger()
      expect(html()).toMatchSnapshot()
    })

    it('is the same hook', () => {
      let i = 0
      let hook!: Hook
      const hooks: Set<Hook> = new Set()
      const Foo = () => {
        hook = current.hook!
        hooks.add(hook)
        return <>{i++}</>
      }
      render(<Foo />, c)
      expect(html()).toMatchSnapshot()
      hook.trigger()
      expect(html()).toMatchSnapshot()
      hook.trigger()
      expect(html()).toMatchSnapshot()
      hook.trigger()
      expect(html()).toMatchSnapshot()
      expect(hooks.size).toEqual(1)
    })

    it('works with nested', () => {
      let i = 0
      let hook!: Hook
      const hooks: Set<Hook> = new Set()
      const Bar = () => {
        hook = current.hook!
        hooks.add(hook)
        return <p>{i++}</p>
      }
      const Foo = () => (
        <div>
          <Bar />
        </div>
      )
      render(<Foo />, c)
      expect(html()).toMatchSnapshot()
      hook.trigger()
      expect(html()).toMatchSnapshot()
      hook.trigger()
      expect(html()).toMatchSnapshot()
      hook.trigger()
      expect(html()).toMatchSnapshot()
      expect(hooks.size).toEqual(1)
    })
  })

  describe('keyed list', () => {
    it('create', () => {
      const Foo = ({ key }: { key: number }) => <li>{key}</li>
      render(
        <>
          {[1, 2, 3].map(x => (
            <Foo key={x} />
          ))}
        </>,
        c
      )
      expect(html()).toMatchSnapshot()
    })

    it('change deep', () => {
      let i = 0
      render(
        <>
          {[1, 2, 3].map(x => (
            <li key={x}>
              {i++ % 2 === 0 && (
                <>
                  <span>{i}</span>
                  <span>{i}</span>
                  <span>{i}</span>
                </>
              )}
            </li>
          ))}
        </>,
        c
      )
      expect(html()).toMatchSnapshot()
      render(
        <>
          {[1, 2, 3].map(x => (
            <li key={x}>
              {i++ % 2 === 0 && (
                <>
                  <span>{i}</span>
                  <span>{i}</span>
                  <span>{i}</span>
                </>
              )}
            </li>
          ))}
        </>,
        c
      )
      expect(html()).toMatchSnapshot()
    })

    it('rearrange', () => {
      const Foo = ({ key }: { key: number }) => <li>{key}</li>
      render(
        <>
          {[1, 2, 3].map(x => (
            <Foo key={x} />
          ))}
        </>,
        c
      )
      expect(html()).toMatchSnapshot()
      const lis = c.querySelectorAll('li')
      render(
        <>
          {[3, 1, 2].map(x => (
            <Foo key={x} />
          ))}
        </>,
        c
      )
      expect(html()).toMatchSnapshot()
      const res = c.querySelectorAll('li')
      expect(res[0] === lis[2]).toEqual(true)
      expect(res[0]).toBe(lis[2])
      expect(res[1]).toBe(lis[0])
      expect(res[2]).toBe(lis[1])
    })

    it('insert', () => {
      const Foo = ({ key }: { key: number }) => <li>{key}</li>
      render(
        <>
          {[1, 2, 3].map(x => (
            <Foo key={x} />
          ))}
        </>,
        c
      )
      expect(html()).toMatchSnapshot()
      const lis = c.querySelectorAll('li')
      render(
        <>
          {[0, 1, 2, 3].map(x => (
            <Foo key={x} />
          ))}
        </>,
        c
      )
      expect(html()).toMatchSnapshot()
      const res = c.querySelectorAll('li')
      expect(res[1]).toBe(lis[0])
      expect(res[2]).toBe(lis[1])
      expect(res[3]).toBe(lis[2])
    })

    it('remove', () => {
      const Foo = ({ key }: { key: number }) => <li>{key}</li>
      render(
        <>
          {[1, 2, 3].map(x => (
            <Foo key={x} />
          ))}
        </>,
        c
      )
      expect(html()).toMatchSnapshot()
      const lis = c.querySelectorAll('li')
      expect(lis.length).toEqual(3)
      render(
        <>
          {[1, 3].map(x => (
            <Foo key={x} />
          ))}
        </>,
        c
      )
      expect(html()).toMatchSnapshot()
      const res = c.querySelectorAll('li')
      expect(res.length).toEqual(2)
      expect(res[0]).toBe(lis[0])
      expect(res[1]).toBe(lis[2])
    })
  })

  describe('svg', () => {
    it('svg', () => {
      render(<svg></svg>, c)
      expect(html()).toMatchSnapshot()
    })

    it('uses correct namespace', () => {
      render(<svg viewBox="0 0 10 10"></svg>, c)
      expect(html()).toMatchSnapshot()
    })

    it('svg then foreignObject', () => {
      render(
        <svg viewBox="0 0 10 10">
          <path pathLength="90" />
          <foreignObject>
            {/* eslint-disable-next-line @typescript-eslint/ban-ts-comment */}
            {/* @ts-ignore */}
            <input autoFocus />
          </foreignObject>
        </svg>,
        c
      )
      expect(html()).toMatchSnapshot()
    })
  })
})
