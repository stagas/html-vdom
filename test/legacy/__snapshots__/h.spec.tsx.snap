// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render create Element 1`] = `"<div></div>"`;

exports[`render create Element with number child 1`] = `"<div>123</div>"`;

exports[`render create Element with text child 1`] = `"<div>hello</div>"`;

exports[`render create Fragment 1`] = `"hello world"`;

exports[`render create Function 1`] = `"hello"`;

exports[`render create Function returning Function 1`] = `"bar"`;

exports[`render create attribute: id 1`] = `"<div id=\\"foo\\"></div>"`;

exports[`render hook is the same hook 1`] = `"0"`;

exports[`render hook is the same hook 2`] = `"1"`;

exports[`render hook is the same hook 3`] = `"2"`;

exports[`render hook is the same hook 4`] = `"3"`;

exports[`render hook updates 1`] = `"0"`;

exports[`render hook updates 2`] = `"1"`;

exports[`render hook works with nested 1`] = `"<div><p>0</p></div>"`;

exports[`render hook works with nested 2`] = `"<div><p>1</p></div>"`;

exports[`render hook works with nested 3`] = `"<div><p>2</p></div>"`;

exports[`render hook works with nested 4`] = `"<div><p>3</p></div>"`;

exports[`render keyed list change deep 1`] = `"<li><span>1</span><span>1</span><span>1</span></li><li></li><li><span>3</span><span>3</span><span>3</span></li>"`;

exports[`render keyed list change deep 2`] = `"<li></li><li><span>5</span><span>5</span><span>5</span></li><li></li>"`;

exports[`render keyed list create 1`] = `"<li>1</li><li>2</li><li>3</li>"`;

exports[`render keyed list insert 1`] = `"<li>1</li><li>2</li><li>3</li>"`;

exports[`render keyed list insert 2`] = `"<li>0</li><li>1</li><li>2</li><li>3</li>"`;

exports[`render keyed list rearrange 1`] = `"<li>1</li><li>2</li><li>3</li>"`;

exports[`render keyed list rearrange 2`] = `"<li>3</li><li>1</li><li>2</li>"`;

exports[`render keyed list remove 1`] = `"<li>1</li><li>2</li><li>3</li>"`;

exports[`render keyed list remove 2`] = `"<li>1</li><li>3</li>"`;

exports[`render svg svg 1`] = `"<svg></svg>"`;

exports[`render svg svg then foreignObject 1`] = `"<svg viewBox=\\"0 0 10 10\\"><path pathLength=\\"90\\"></path><foreignObject><input autofocus=\\"\\"></foreignObject></svg>"`;

exports[`render svg uses correct namespace 1`] = `"<svg viewBox=\\"0 0 10 10\\"></svg>"`;

exports[`render update Element 1`] = `"<div></div>"`;

exports[`render update Element 2`] = `"<p></p>"`;

exports[`render update Element with number child 1`] = `"<div>123</div>"`;

exports[`render update Element with number child 2`] = `"<div>456</div>"`;

exports[`render update Element with text child 1`] = `"<div>hello</div>"`;

exports[`render update Element with text child 2`] = `"<div>world</div>"`;

exports[`render update Fragment 1`] = `"hello world"`;

exports[`render update Fragment 2`] = `"something else"`;

exports[`render update Function 1`] = `"0"`;

exports[`render update Function 2`] = `"1"`;

exports[`render update Function returning Function 1`] = `"0"`;

exports[`render update Function returning Function 2`] = `"1"`;

exports[`render update only updates what changed 1`] = `"<div><p>0</p></div>"`;

exports[`render update only updates what changed 2`] = `"<div><p>1</p></div>"`;
